### PR TITLE
Destructuring parameters in function/constructor types

### LINKED ISSUE #6840: Binding patterns in parameter lists of function types
I believe that the following two variable declarations are consistent with the specifications ([3.9.2.2](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#3922-parameter-list)).  They define two functions of a single parameter, which is a list of three things.  Notice that a binding pattern is used for destructuring this list.

```TypeScript
var y: ([a, b, c]) => any;
var z: ([a, b, c] : number[]) => any;
```

Both declarations are disallowed by the implementation.  The error messages suggest that `[a, b, c]` is immediately assumed to be a (parenthesized) tuple type, as would be the case in:

```TypeScript
interface a {};
interface b {};
interface c {};

var x: ([a, b, c]);
```

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #6840.

### CODE REVIEW

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -2450,10 +2450,27 @@ namespace ts {
...
+                // Skip modifiers
+                parseModifiers();
+            }
+            if (isIdentifier()) {
+                nextToken();
+                return true;
+            }
+            if (token === SyntaxKind.OpenBracketToken || token === SyntaxKind.OpenBraceToken) {
+                // Return true if we can parse an array or object binding pattern with no errors
+                const count = parseDiagnostics.length;
```
Comment (sandersn) [inline src/compiler/parser.ts]: maybe a name like previousCount or previousErrorCount?

File: src/compiler/parser.ts [hunk 2]
```diff
@@ -2450,10 +2450,27 @@ namespace ts {
...
                 return true;
             }
-
             return token === SyntaxKind.OpenParenToken && lookAhead(isUnambiguouslyStartOfFunctionType);
         }
 
+        function skipParameterStart(): boolean {
+            if (isModifierKind(token)) {
+                // Skip modifiers
+                parseModifiers();
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: Shouldn't there be errors on the modifiers?

File: src/compiler/parser.ts [hunk 3]
```diff
@@ -2450,10 +2450,27 @@ namespace ts {
...
+            }
+            if (isIdentifier()) {
+                nextToken();
+                return true;
+            }
+            if (token === SyntaxKind.OpenBracketToken || token === SyntaxKind.OpenBraceToken) {
+                // Return true if we can parse an array or object binding pattern with no errors
+                const count = parseDiagnostics.length;
+                parseIdentifierOrPattern();
+                return count === parseDiagnostics.length;
```
Comment (DanielRosenwasser) [inline src/compiler/parser.ts]: Why do it this way? Why not just `tryParse`?

Comment (ahejlsberg) [inline src/compiler/parser.ts]: We're already in a `lookAhead` call so no point in doing speculative parsing again. Also, `tryParse` doesn't check whether errors were generated.

File: src/compiler/parser.ts (hunk 1 above)
Comment (ahejlsberg) [inline src/compiler/parser.ts]: Sure.

**Comment by sandersn** [comment]
:+1:

**Pull request merged**